From d428fd49ca22e8408f8c84a99b89c43f8b6ad9a4 Mon Sep 17 00:00:00 2001
From: Sai Vemuri <svemuri@redhat.com>
Date: Thu, 11 Sep 2014 22:43:13 -0400
Subject: [infiniband] cxgb4: Fix over-dereference when terminating

Message-id: <1410475447-94392-55-git-send-email-svemuri@redhat.com>
Patchwork-id: 93826
O-Subject: [RHEL7.1 PATCH BZ 1124947 054/108] RDMA/cxgb4: Fix over-dereference when terminating
Bugzilla: 1124947
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Need to get the endpoint reference before calling rdma_fini(), which
might fail causing us to not get the reference.

Signed-off-by: Steve Wise <swise@opengridcomputing.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
(cherry picked from commit 1d1ca9b4fdde07325d263f7a75379527b1281f52)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/cxgb4/qp.c b/drivers/infiniband/hw/cxgb4/qp.c
index 5a7d368..7b5114c 100644
--- a/drivers/infiniband/hw/cxgb4/qp.c
+++ b/drivers/infiniband/hw/cxgb4/qp.c
@@ -1389,6 +1389,7 @@ int c4iw_modify_qp(struct c4iw_dev *rhp, struct c4iw_qp *qhp,
    qhp->attr.ecode = attrs->ecode;
    ep = qhp->ep;
    disconnect = 1;
+   c4iw_get_ep(&qhp->ep->com);
    if (!internal)
     terminate = 1;
    else {
@@ -1396,7 +1397,6 @@ int c4iw_modify_qp(struct c4iw_dev *rhp, struct c4iw_qp *qhp,
     if (ret)
      goto err;
    }
-   c4iw_get_ep(&qhp->ep->com);
    break;
   case C4IW_QP_STATE_ERROR:
    t4_set_wq_in_error(&qhp->wq);
-- 
1.7.1